features/cellular/easy_cellular/CellularConnectionFSM.cpp

```diff
@@ -268,223 +334,197 @@ nsapi_error_t CellularConnectionFSM::continue_to_state(CellularState state)
     return NSAPI_ERROR_OK;
 }
 
-void CellularConnectionFSM::event()
+void CellularConnectionFSM::enter_to_state(CellularState state)
+{
+    _next_state = state;
+    _retry_count = 0;
+}
+
+void CellularConnectionFSM::retry_state_or_fail()
+{
+    if (++_retry_count < MAX_RETRY_ARRAY_SIZE) {
+        tr_debug("Retry State %s, retry %d/%d", get_state_string(_state), _retry_count, MAX_RETRY_ARRAY_SIZE);
+        _event_timeout = _retry_timeout_array[_retry_count];
+    } else {
+        report_failure(get_state_string(_state));
+        return;
+    }
+}
+
+void CellularConnectionFSM::state_init()
+{
+    _event_timeout = _start_time;
+    tr_info("Init state, waiting %d ms before POWER state)", _start_time);
+    enter_to_state(STATE_POWER_ON);
+}
+
+void CellularConnectionFSM::state_power_on()
 {
-    nsapi_error_t err;
-    int event_timeout = -1;
+    _cellularDevice->set_timeout(TIMEOUT_POWER_ON);
+    tr_info("Cellular power ON (timeout %d ms)", TIMEOUT_POWER_ON);
+    if (power_on()) {
+        enter_to_state(STATE_DEVICE_READY);
+    } else {
+        // retry to power on device
+        retry_state_or_fail();
+    }
+}
+
+bool CellularConnectionFSM::device_ready()
+{
+    tr_info("Cellular device ready");
+    if (_event_status_cb) {
+        _event_status_cb((nsapi_event_t)CellularDeviceReady, 0);
+    }
+
+    _power->remove_device_ready_urc_cb(mbed::callback(this, &CellularConnectionFSM::ready_urc_cb));
+
+    bool success = false;
+    for (int type = 0; type < CellularNetwork::C_MAX; type++) {
+        if (!_network->set_registration_urc((CellularNetwork::RegistrationType)type, true)) {
+             success = true;
+        }
+    }
+    if (!success) {
+        tr_error("Failed to set any URC's for registration");
+        report_failure(get_state_string(_state));
+        return false;
+    }
+
+    return true;
+}
 
+void CellularConnectionFSM::state_device_ready()
+{
+    _cellularDevice->set_timeout(TIMEOUT_POWER_ON);
+    if (_power->set_at_mode() == NSAPI_ERROR_OK) {
+        if (device_ready()) {
+            enter_to_state(STATE_SIM_PIN);
+        }
+    } else {
+        if (_retry_count == 0) {
+            (void)_power->set_device_ready_urc_cb(mbed::callback(this, &CellularConnectionFSM::ready_urc_cb));
+        }
+        retry_state_or_fail();
+    }
+}
+
+void CellularConnectionFSM::state_sim_pin()
+{
+    _cellularDevice->set_timeout(TIMEOUT_SIM_PIN);
+    tr_info("Sim state (timeout %d ms)", TIMEOUT_SIM_PIN);
+    if (open_sim()) {
+        enter_to_state(STATE_REGISTERING_NETWORK);
+    } else {
+        retry_state_or_fail();
+    }
+}
+
+void CellularConnectionFSM::state_registering()
+{
+    _cellularDevice->set_timeout(TIMEOUT_NETWORK);
+
+    if (is_registered()) {
```

Comment:
Could we trigger COPS=0 once and then rely on CREGs to report connection status due to COPS=0 will block during while AT commands will fail.